mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs

```diff
@@ -0,0 +1,1494 @@
+            ticker: ticker.clone(),
+        };
+        let walletdb_blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+        let height = u32::from(height);
+        let index_keys = MultiIndex::new(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
+            .with_value(&ticker)?
+            .with_value(num_to_bigint!(height)?)?;
+        walletdb_blocks_table
+            .replace_item_by_unique_multi_index(index_keys, &block)
+            .await?;
+
+        Ok(())
+    }
+}
+
+impl WalletIndexedDb {
+    pub async fn insert_block(
+        &self,
+        block_height: BlockHeight,
+        block_hash: BlockHash,
+        block_time: u32,
+        commitment_tree: &CommitmentTree<Node>,
+    ) -> MmResult<(), ZcoinStorageError> {
+        let ticker = self.ticker.clone();
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let walletdb_blocks_table = db_transaction.table::<WalletDbBlocksTable>().await?;
+
+        let mut encoded_tree = Vec::new();
+        commitment_tree.write(&mut encoded_tree).unwrap();
+
+        let hash = &block_hash.0[..];
+        let block = WalletDbBlocksTable {
+            height: u32::from(block_height),
+            hash: hash.to_vec(),
+            time: block_time,
+            sapling_tree: encoded_tree,
+            ticker: ticker.clone(),
+        };
+
+        let index_keys = MultiIndex::new(WalletDbBlocksTable::TICKER_HEIGHT_INDEX)
+            .with_value(&ticker)?
+            .with_value(u32::from(block_height))?;
+
+        Ok(walletdb_blocks_table
+            .replace_item_by_unique_multi_index(index_keys, &block)
+            .await
+            .map(|_| ())?)
+    }
+
+    pub async fn get_balance(&self, account: AccountId) -> MmResult<Amount, ZcoinStorageError> {
+        let ticker = self.ticker.clone();
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let rec_note_table = db_transaction.table::<WalletDbReceivedNotesTable>().await?;
+
+        let index_keys = MultiIndex::new(WalletDbReceivedNotesTable::TICKER_ACCOUNT_INDEX)
+            .with_value(&ticker)?
+            .with_value(account.0.to_bigint().unwrap())?;
+        let maybe_notes = rec_note_table.get_items_by_multi_index(index_keys).await?;
+
+        let tx_table = db_transaction.table::<WalletDbTransactionsTable>().await?;
+        let txs = tx_table.get_items("ticker", &ticker).await?;
+
+        let mut balance: i64 = 0;
+        for (_, note) in &maybe_notes {
+            let spent = &note.spent;
+            let value = &note.value;
+
+            for (tx_id, tx) in &txs {
+                if *tx_id == note.tx && spent.is_none() && tx.block.is_some() {
+                    let value_i64 = value.to_i64().expect("BigInt is too large to fit in an i64");
+                    balance += value_i64;
+                }
+            }
+        }
```

Comment:
We should be able to achieve this without mutating `balance` using filters and maps on `maybe_notes`.

Reply:
done